tilings/algorithms/gridded_perm_reduction.py

```diff
@@ -0,0 +1,305 @@
+        requirements: Iterable[Tuple[GriddedPerm, ...]],
+    ) -> List[Tuple[GriddedPerm, ...]]:
+        """
+        Add factors of requirements as size one requirements, removing them
+        from each of the gridded permutations in the requirement it is
+        contained in.
+        """
+        res: List[Tuple[GriddedPerm, ...]] = list()
+        for requirement in requirements:
+            if not requirement:
+                # If req is empty, then this requirement can't be satisfied, so
+                # return it and mark obstructions and requirements as empty.
+                return [tuple()]
+            # If any gridded permutation in list is empty then you vacuously
+            # contain this requirement
+            if all(requirement):
+                factors = GriddedPermReduction.factors(requirement)
+                if len(factors) == 0 or (len(factors) == 1 and len(requirement) == 1):
+                    # if there are no factors in the intersection, or it is just
+                    # the same req as the first, we do nothing and add the original
+                    res.append(requirement)
+                    continue
+                # add each of the factors as a single requirement, and then remove
+                # these from each of the other requirements in the list
+                remaining_cells = set(c for gp in requirement for c in gp.pos) - set(
+                    c for gp in factors for c in gp.pos
+                )
+                for factor in factors:
+                    res.append((factor,))
+                res.append(
+                    tuple(
+                        gp.get_gridded_perm_in_cells(remaining_cells)
+                        for gp in requirement
+                    )
+                )
+        return res
+
+    # if a gridded perm G in a requirment has a factor that appears in EVERY gridded
+    # perm of some other requirement, then that factor can be removed from G
+    # [subrequirement inferral]
+    # NOTE: This handles the following case:
+    #   requirement R2 makes requirement R1 redundant if:
+    #   For all G in R2, there exists H in R1 such that H <= G
+    @cssmethodtimer("GriddedPermReduction.cleaned_requirements")
+    def cleaned_requirements(
+        self, requirements: List[Tuple[GriddedPerm, ...]]
+    ) -> List[Tuple[GriddedPerm, ...]]:
+        """
+        Remove the factors of a gridded permutation in a requirement if it is
+        implied by another requirement.
+        """
+        for idx, requirement in enumerate(requirements):
+            if not all(requirement):
+                continue
+            newgps: List[GriddedPerm] = []
+            for gp in requirement:
+                cells: List[Cell] = []
+                for f in gp.factors():
+                    if not self._griddedperm_implied_by_some_requirement(
+                        f,
+                        (
+                            other_requirement
```

Comment:
Can we just replace this big thing by
`requirements[:idx] + requirements[idx+1:]`

Reply:
Or if your concern about creating new list we can also use itertools.islice

Reply:
I'm not sure I understand this comment. This function shrinks GPs in requirements by removing some factors.

Reply:
I meant that the second argument of `_griddedperm_implied_by_some_requirement` could be replaced  by this which is much shorted.

Reply:
Oh, I see. I think itertools.islice would be okay. @christianbean?

Reply:
That is the best option

Reply:
Fixed.